tests/test_utils.py

```diff
@@ -178,3 +183,35 @@ def test_async_barrier_reset(self):
         assert barrier.count == 5
         barrier.reset()
         assert barrier.count == 0
+
+
+@pytest.mark.asyncio
+async def test_encoded_text_write():
+    # 1. Get temporary filename
+    with NamedTemporaryFile() as temp_file:
+        file_name = temp_file.name
+
+    # 2. Generate random string
+    init_str = (''.join(choice(ascii_uppercase) for i in range(100)))
+
+    # 3. Write chuncked decoded string into file
+    async with AsyncFileWriter(
+            current_loop(),
+            file_name,
+            access_mode='w+',
+            decode=codecs.decode) as file_writer:
+        size = 300
+        if file_writer._max_chunks is None or file_writer._max_chunks >= 5:
+            step = 60
+        else:
+            step = 300 / file_writer._max_chunks
```

Comment:
Isn't this a leak of abstraction?
Your test should use the public interface only, unless you have to mock underlying external resource accesses.
`AsyncFileWriter` should work (does not hang) regardless of what `max_chunks` value is, and the callers should have no need to change their codes depending on the hidden details.

Reply:
For instance, what if I set `max_chunks=500` here?
Your fix only applies to your specific test code instance.

Reply:
Fair enough, but then I'm kinda questioning the necessity of checking that at all. Having max_chunks is just an implementation detail (in manager it limits the size of the queue for whatever reason). In my opinion, it is reasonable to leave that as a responsibility for the user of this context manager, or, if it is appropriate for the manager side code, omit this parameter (but then again, in case someone would need to limit the queue, there will be no option to do that). Wanna hear your opinion on this.

Reply:
The point of limiting the maximum number of chunks is to limit the memory usage (buffering size) for large-file uploads.
Your current implementation breaks this design because all chunks are buffered at once and all chunks are written at once, beacuse `_write()` method is called at `__aexit__` while the caller invokes `write()` method before it.
You should spawn the executor task at `__aenter__` and await until its termination at `__aexit__`.